runtime/compiler/optimizer/SequentialStoreSimplifier.cpp

```diff
@@ -677,621 +684,2233 @@ TR::Node* getALoadReferenceForSeqLoad(TR::Node* rootNode, int32_t totalBytes, in
       return currentTreeTop;
+      }
 
-   int32_t numBytes = 0;
-   TR::Node* currentNode = currentTreeTop->getNode();
-   TR::Node* prevNode = NULL;
-   TR::Node* rootNode;
+   if (trace) traceMsg(comp, "Sequential Load Simplification Candidate - rootNode: %p, %s\n", rootNode, littleEndianLoad ? "Little Endian Load" : "Big Endian Load");
+   if (trace) traceMsg(comp, "Sequential Load Simplification Candidate - rootNode: %p, newConvertChildNode: %p, newLoadChildNode: %p\n", rootNode, newConvertChildNode, newLoadChildNode);
 
-   //checking the number of bytes
-   while ( !((currentNode->getFirstChild() == ibloadNode) && (currentNode->getOpCodeValue() == TR::bu2i)) )
+   /*
+    * Check the endianess of the platform and the endianess of the byte array.
+    * If they don't match, byteswaps will be needed.
+    */
+   if (littleEndianLoad && !comp->target().cpu.isLittleEndian())
+      {
+      if (!comp->cg()->supportsByteswap())
+         {
+         if (trace) traceMsg(comp, "Little endian load on big endian target without byteswap is not supported. rootNode: %p\n", rootNode);
+         return currentTreeTop;
+         }
+      swapBytes = true;
+      }
+   else if (!littleEndianLoad && comp->target().cpu.isLittleEndian())
       {
-      if ( (currentNode->getOpCodeValue() == TR::iadd) || (currentNode->getOpCodeValue() == TR::ior))
+      if (!comp->cg()->supportsByteswap())
          {
-         numBytes++;
-         if (numBytes==1)
-            rootNode=prevNode;
+         if (trace) traceMsg(comp, "Big endian load on little endian target without byteswap is not supported. rootNode: %p\n", rootNode);
+         return currentTreeTop;
          }
-      else if (isValidSeqLoadIMul(comp, currentNode))
-         numBytes++;
-      else
-         numBytes = 0;
-      prevNode = currentNode;
-      currentNode = currentNode->getFirstChild();
+      swapBytes = true;
       }
 
-   if (numBytes == 1)
+   if (!performTransformation(comp, "%sReducing sequential loads\n", OPT_DETAILS))
       {
-      dumpOptDetails(comp, " Sequential Load of 1 byte not worth reducing\n");
       return currentTreeTop;
       }
-   else if ( numBytes != 2 && numBytes != 3 && numBytes != 4 )
+
+   /* All checks have passed so perform the transformation. */
+   traceMsg(comp, "Sequential Load reduced at rootNode: %p\n", rootNode);
+
+   disconnectedNode1 = rootNode->getFirstChild();
+   disconnectedNode2 = rootNode->getSecondChild();
+
+   if (((4 == byteCount) && !swapBytes && !is64bitResult) ||
+       ((8 == byteCount) && !swapBytes))
       {
-      dumpOptDetails(comp, " Sequential Load of size other than 2/3/4 not reducible\n");
-      return currentTreeTop;
-      }
+      if (4 == byteCount)
+         {
+         /* byteCount is 4, final result is an int, no byteswaps */
+         if (trace) traceMsg(comp, "Recreating rootNode (%p) as iloadi.\n", rootNode);
+         TR::Node::recreateWithSymRef(rootNode, TR::iloadi, newLoadChildNode->getSymbolReference());
```

Comment:
Will we have an iloadi node with a symbol reference of type byte(Int8) ? I think this can lead to some complications in the optimizer and the codegens.

Reply:
I'm not sure if this is a problem. I will look into it.

Reply:
In the latest changes I changed the symref to being a generic int array shadow symref.